core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java

```diff
@@ -518,19 +520,52 @@ public int getNumRecords() {
 
     public long spill() throws IOException {
```

Comment:
low-level code is hard to review as nobody remembers all the details all the time. Can you briefly explain the spill process to help people review?

Reply:
Before [SPARK-14851](https://issues.apache.org/jira/browse/SPARK-14851), ```UnsafeInMemorySorter.getSortedIterator()```, the returned type is ```UnsafeInMemorySorter.SortedIterator```.

Now, when the radio sort is turned on and there are some empty values, the return type is ```UnsafeExternalSorter.ChainedIterator(UnsafeInMemorySorter.SortedIterator)```.
In ```UnsafeExternalSorter.SpillableIterator#spill```, this situation cannot be spilled.


When ```UnsafeExternalSorter#getSortedIterator``` is called, part of the last written data is saved in memory.

When the task needs more execution memory, the spill fails.

TaskMemoryManager#acquireExecutionMemory->UnsafeExternalSorter#spill->UnsafeExternalSorter.SpillableIterator#spill